std/experimental/checkedint.d

```diff
@@ -0,0 +1,1626 @@
+
+    If a cast to a floating-point type is requested and `Hook` defines
+    `onBadCast`, the cast is verified by ensuring $(D representation == cast(T)
+    U(representation)). If that is not `true`, `hook.onBadCast!U(payload)` is
+    returned.
+
+    If a cast to an integral type is requested and `Hook` defines `onBadCast`,
+    the cast is verified by ensuring `representation` and $(D cast(U)
+    representation) are the same arithmetic number. (Note that `int(-1)` and
+    `uint(1)` are different values arithmetically although they have the same
+    bitwise representation and compare equal by language rules.) If the numbers
+    are not arithmetically equal, `hook.onBadCast!U(payload)` is returned.
+
+    */
+    U opCast(U)()
+    if (isIntegral!U || isFloatingPoint!U || is(U == bool))
+    {
+        static if (hasMember!(Hook, "hookOpCast"))
+        {
+            return hook.hookOpCast!U(payload);
+        }
+        else static if (is(U == bool))
+        {
+            return payload != 0;
+        }
+        else static if (valueConvertible!(T, U))
+        {
+            return payload;
+        }
+        // may lose bits or precision
+        else static if (!hasMember!(Hook, "onBadCast"))
+        {
+            return cast(U) payload;
+        }
+        else
+        {
+            if (isUnsigned!T || !isUnsigned!U ||
+                    T.sizeof > U.sizeof || payload >= 0)
+            {
+                auto result = cast(U) payload;
+                // If signedness is different, we need additional checks
+                if (result == payload &&
+                        (!isUnsigned!T || isUnsigned!U || result >= 0))
+                    return result;
+            }
+            return hook.onBadCast!U(payload);
+        }
+    }
+    ///
+    unittest
+    {
+        assert(cast(uint) Checked!int(42) == 42);
+        assert(cast(uint) Checked!(int, WithNaN)(-42) == uint.max);
+    }
+
+    // opEquals
+    /**
+    Compares `this` against `rhs` for equality. If `Hook` defines
+    `hookOpEquals`, the function forwards to $(D
+    hook.hookOpEquals(representation, rhs)). Otherwise, the result of the
+    built-in operation $(D payload == rhs) is returned.
+
+    If `U` is an instance of `Checked`
+    */
+    bool opEquals(U)(U rhs)
+    if (isIntegral!U || isFloatingPoint!U || is(U == bool) ||
+        is(U == Checked!(V, W), V, W) && is(typeof(this == rhs.payload)))
+    {
+        static if (is(U == Checked!(V, W), V, W))
+        {
+            alias R = typeof(payload + rhs.payload);
+            static if (is(Hook == W))
+            {
+                // Use the lhs hook if there
+                return this == rhs.payload;
+            }
+            else static if (valueConvertible!(T, R) && valueConvertible!(V, R))
+            {
+                return payload == rhs.payload;
+            }
+            else static if (hasMember!(Hook, "hookOpEquals"))
```

Comment:
> Would there be a case when a hook wants to intercept the "correct" comparisons?

I don't know, but if I was writing my own `hookOpEquals` I would naively expect it to get called in every case, instead of just some. If you choose to keep it as is, make sure that the documentation for the `Hook` API clearly explains the circumstances in which it will be called.